File: 0020-valid-parentheses/0020-valid-parentheses.swift
class Solution {
    func isValid(_ s: String) -> Bool {
        var stack = [Character]()
        
        for letter in Array(s) {
            if letter == "(" || letter == "{" || letter == "[" {
                stack.append(letter)  
            } else if letter == ")" {
                if !stack.isEmpty && stack[stack.endIndex - 1] == "(" {
                    stack.removeLast()
                } else {
                    return false
                }
            } else if letter == "}" {
                if !stack.isEmpty && stack[stack.endIndex - 1] == "{" {
                    stack.removeLast()
                } else {
                    return false
                }
            } else if letter == "]" {
                if !stack.isEmpty && stack[stack.endIndex - 1] == "[" {
                    stack.removeLast()
                } else {
                    return false
                }
            }
        }
        
        return stack.isEmpty
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.ts
function isValid(s: string): boolean {
    let stack: string[] = [];

    for (const letter of s) {
        if (stack.length > 0) {
            const top = stack[stack.length - 1];

            if (top === "(" && letter === ")") {
                stack.pop();
            } else if (top === "{" && letter === "}") {
                stack.pop();
            } else if (top === "[" && letter === "]") {
                stack.pop();
            } else {
                stack.push(letter);
            }
        } else {
            stack.push(letter);
        }
    }

    return stack.length === 0;
};